model/dao/ProjectDAO/PostgreSQLProjectDAO.php

```diff
@@ -594,36 +595,58 @@ public function update(ProjectVO $projectVO) {
-                DBPostgres::checkNull($projectVO->getAreaId()) . ", " .
-                DBPostgres::checkNull($projectVO->getCustomerId()) . ", " .
-                DBPostgres::checkStringNull($projectVO->getType()) . ", " .
-                DBPostgres::checkStringNull($projectVO->getDescription()) . ", " .
-                DBPostgres::checkNull($projectVO->getMovedHours()) . ", " .
-                DBPostgres::checkStringNull($projectVO->getSchedType()) .")";
-
-        $res = pg_query($this->connect, $sql);
-
-        if ($res == NULL) throw new SQLQueryErrorException(pg_last_error());
-
-        $projectVO->setId(DBPostgres::getId($this->connect, "project_id_seq"));
-
-        $affectedRows = pg_affected_rows($res);
-
-        return $affectedRows;
+        $result = new OperationResult(true);
+
+        $sql = "INSERT INTO project (activation, init, _end, invoice, est_hours, 
+        areaid, customerid, type, description, moved_hours, sched_type) VALUES(
+            :activation, :init, :end, :invoice, :est_hours,
+        :areaid, :customerid, :type, :description, :moved_hours, :sched_type)";
+
+        try {
+            $statement = $this->pdo->prepare($sql);
+            $statement->bindValue(":activation", $projectVO->getActivation(), PDO::PARAM_BOOL);
+            $statement->bindValue(":init", $projectVO->getInit(), PDO::PARAM_STR);
+            $statement->bindValue(":end", $projectVO->getEnd(), PDO::PARAM_STR);
+            $statement->bindValue(":invoice", $projectVO->getInvoice(), PDO::PARAM_STR);
+            $statement->bindValue(":est_hours", $projectVO->getEstHours(), PDO::PARAM_STR);
+            $statement->bindValue(":areaid", $projectVO->getAreaId(), PDO::PARAM_INT);
+            $statement->bindValue(":customerid", $projectVO->getCustomerId(), PDO::PARAM_INT);
+            $statement->bindValue(":type", $projectVO->getType(), PDO::PARAM_STR);
+            $statement->bindValue(":description", $projectVO->getDescription(), PDO::PARAM_STR);
+            $statement->bindValue(":moved_hours", $projectVO->getMovedHours(), PDO::PARAM_STR);
+            $statement->bindValue(":sched_type", $projectVO->getSchedType(), PDO::PARAM_STR);
+            $statement->execute();
+
+            $projectVO->setId($this->pdo->lastInsertId('project_id_seq'));
+
+            $result->setMessage('Project created successfully.');
+            $result->setResponseCode(201);
+        }
+        catch (PDOException $ex) {
+            //make sure to log the error as a failure, but return the OperationResult object
+            //successfully so that front end can see error and fail gracefully
+            $errorMessage = $ex->getMessage();
+            error_log('Project creation failed: ' . $errorMessage);
+            $result->setErrorNumber($ex->getCode());
+            $resultMessage = "Project creation failed: \n";
+            if(strpos($errorMessage, "Foreign key violation")){
```

Comment:
A "not null" error is also possible, e.g. when you forget to set the area:
```
[Mon Jan 16 10:13:02 2023] Project creation failed: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "areaid" of relation "project" violates not-null constraint
```

Reply:
Null customer is technically allowed although the UI tries to prevent it, we made a mess with that :shrug: :sweat_smile: